core/dbt/parser/manifest.py

```diff
@@ -599,6 +652,173 @@ def write_manifest_for_partial_parse(self):
         except Exception:
             raise
 
+    def write_artifacts(self):
+        # The manifest.json is written out in a task, so we're not writing it here
+
+        # build publication metadata
+        metadata = PublicationMetadata(
+            adapter_type=self.root_project.credentials.type,
+            quoting=self.root_project.quoting,
+        )
+
+        # get a list of public model ids first so it can be used in constructing dependencies
+        public_model_ids = []
+        for node in self.manifest.nodes.values():
+            if node.resource_type == NodeType.Model and node.access == AccessType.Public:
+                public_model_ids.append(node.unique_id)
+
+        set_of_public_unique_ids = set(public_model_ids)
+
+        # Get the Graph object from the Linker
+        from dbt.compilation import Linker
+
+        linker = Linker()
+        graph = linker.get_graph(self.manifest)
+
+        public_models = {}
+        for unique_id in public_model_ids:
+            model = self.manifest.nodes[unique_id]
+            # public_node_dependencies is the intersection of all parent nodes plus public nodes
+            public_node_dependencies = []
+            parents: set = graph.select_parents({unique_id})
+            public_node_dependencies = parents.intersection(set_of_public_unique_ids)
+
+            public_model = PublicModel(
+                name=model.name,
+                package_name=model.package_name,
+                unique_id=model.unique_id,
+                relation_name=model.relation_name,
+                version=model.version,
+                latest_version=model.latest_version,
+                public_node_dependencies=list(public_node_dependencies),
+                generated_at=metadata.generated_at,
+            )
+            public_models[unique_id] = public_model
+
+        dependencies = []
+        if self.manifest.project_dependencies:
+            for project in self.manifest.project_dependencies.projects:
```

Comment:
We'll need to include transitive dependencies (available through `self.manifest.publications`) in the publication artifact to [detect project-level cycles](https://github.com/dbt-labs/dbt-core/issues/7468). 

The high-level logic here would be to loop through the available publications in self.manifest.publications, and include their `dependencies` in the dependencies being constructed in addition to including the dependencies in self.manifest.project_dependencies.